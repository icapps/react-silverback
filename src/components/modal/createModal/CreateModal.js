import React from 'react';
import PropTypes from 'prop-types';
import { Modal, Checkbox, BasicInput, Dropdown } from '../../index';
import { strings } from '../../../utils';
import { identifiers } from '../../../constants';
import './createModal.css';
import { regexes } from '../../../constants';

const plus = require('../../../assets/images/plus.svg');

class CreateModal extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      inputs: this.initializeCreateParameters(),
      createPassword: true,
    };
  }

  initializeCreateParameters = () => {
    let inputs = {};
    this.props.createParameters.forEach(item => {
      if (item.type === 'email') {
        inputs[item.id] = { value: '', validation: 'email', isValid: true, errorMessage: '' };
      } else if (item.type === 'text') {
        inputs[item.id] = { value: '', validation: 'text', isValid: true, errorMessage: '' };
      } else if (item.type === 'password') {
        inputs[item.id] = { value: '', validation: this.createPassword ? 'password' : 'none', isValid: true, errorMessage: '' };
      } else if (item.type === 'select' && item.options.length > 0) {
        inputs[item.id] = { value: item.options[0].key, validation: 'none' };
      } else if (item.type === 'boolean') {
        inputs[item.id] = { value: item.value, validation: 'none' };
      }
    });
    return inputs;
  }

  setCreateParameters = () => {
    this.setState({ inputs: this.initializeCreateParameters() });
  }

  create = () => {
    let toValidate = 0;
    let validated = 0;
    let inputs = { ...this.state.inputs };

    for (let key in inputs) {
      if (inputs.hasOwnProperty(key) && inputs[key].validation !== 'none') {
        toValidate++;
        let validation = this.validate(inputs[key].validation, inputs[key].value);
        inputs[key].isValid = validation.isValid;
        inputs[key].errorMessage = validation.errorMessage;
        if (validation.isValid) {
          validated++;
        }
      }
    }

    if (toValidate === validated) {
      for (let key in inputs) {
        if (inputs.hasOwnProperty(key)) {
          inputs[key] = inputs[key].value;
        }
      }
      if (this.state.createPassword) {
        return this.props.create({ ...inputs, password: undefined }, this.state.createPassword);
      }
      return this.props.create(inputs, this.state.createPassword);
    } else {
      this.setState({ inputs });
    }
  }

  handleChange = event => {
    const input = event.target.id.replace('modal-', '');
    this.setState({
      inputs: {
        ...this.state.inputs,
        [input]: {
          ...this.state.inputs[input],
          value: event.target.type === 'checkbox' ? event.target.checked : event.target.value,
          isValid: true,
        },
      },
    });
  };

  handleCreatePassword = () => {
    this.setState({
      createPassword: !this.state.createPassword,
    }, () => {
      this.setState({
        inputs: {
          ...this.state.inputs,
          password: {
            ...this.state.inputs['password'],
            validation: this.state.createPassword ? 'none' : 'password',
            isValid: true,
          },
        },
      });
    });
  }

  renderInput = item => {
    if (item.type === 'boolean') {
      return <Checkbox key={item.id} id={`modal-${item.id}`} text={item.label} value={this.state.inputs[item.id].value} handleChange={this.handleChange} isDisabled={this.props.isPending} />;
    } else if (item.type === 'select' && item.options && this.state.inputs[item.id]) {
      return <Dropdown key={item.id} id={`modal-${item.id}`} label={item.label} value={this.state.inputs[item.id].value} handleChange={this.handleChange} options={item.options} />;
    } else if (item.type === 'password') {
      return (
        <React.Fragment key={item.id}>
          <Checkbox id={identifiers.USER_HAS_TO_SET_PASSWORD} text={strings.USER_HAS_TO_SET_PASSWORD} value={this.state.createPassword} handleChange={this.handleCreatePassword} isDisabled={this.props.isPending} />
          <BasicInput id={`modal-${item.id}`} label={item.label} value={this.state.inputs[item.id].value} isValid={this.state.inputs[item.id].isValid} errorMessage={this.state.inputs[item.id].errorMessage} handleChange={this.handleChange} type={item.type} isDisabled={this.props.isPending || this.state.createPassword} />
        </React.Fragment>
      );
    } else if (this.state.inputs[item.id]) {
      return <BasicInput key={item.id} id={`modal-${item.id}`} label={item.label} value={this.state.inputs[item.id].value} isValid={this.state.inputs[item.id].isValid} errorMessage={this.state.inputs[item.id].errorMessage} handleChange={this.handleChange} type={item.type} isDisabled={this.props.isPending} />;
    }
  }

  validate = (type, value) => {
    if (type === 'email') {
      if (value === '') {
        return {
          isValid: false,
          errorMessage: strings.LOGIN_EMAIL_REQUIRED,
        };
      } else if (!regexes.EMAIL.test(value)) {
        return {
          isValid: false,
          errorMessage: strings.LOGIN_EMAIL_VALIDATION,
        };
      } else {
        return {
          isValid: true,
          errorMessage: '',
        };
      }
    } else if (type === 'password') {
      if (value === '') {
        return {
          isValid: false,
          errorMessage: strings.LOGIN_PASSWORD_REQUIRED,
        };
      } else if (value.length < 6) {
        return {
          isValid: false,
          errorMessage: strings.PASSWORD_LENGTH,
        };
      } else {
        return {
          isValid: true,
          errorMessage: '',
        };
      }
    } else if (type === 'text') {
      if (value === '') {
        return {
          isValid: false,
          errorMessage: strings.FIELD_REQUIRED,
        };
      } else {
        return {
          isValid: true,
          errorMessage: '',
        };
      }
    }
  }

  render() {
    return (
      <Modal
        id={strings.CREATE}
        modalButtonText={this.props.primaryButtonText}
        secondaryButtonText={strings.CANCEL}
        handlePrimaryButton={this.create}
        primaryButtonText={strings.SUBMIT}
        primaryButtonClassName="btn-success"
        modalButtonClassName="btn-success btn-create"
        icon={plus}
        title={this.props.title}
        hasHeader
        isPending={this.props.isPending}
        handleModalButton={this.setCreateParameters}
      >
        <div className="create-modal-input">
          {this.state.inputs && this.props.createParameters.map(item => this.renderInput(item))}
        </div>
      </Modal>
    );
  }
}

CreateModal.propTypes = {
  primaryButtonText: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  createParameters: PropTypes.array.isRequired,
  create: PropTypes.func.isRequired,
  isPending: PropTypes.bool,
  isError: PropTypes.bool,
  errorMessage: PropTypes.string,
};

CreateModal.defaultProps = {
  isPending: false,
  isError: false,
  errorMessage: '',
};

export default CreateModal;
