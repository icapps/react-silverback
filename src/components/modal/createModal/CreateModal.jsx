import React from 'react';
import PropTypes from 'prop-types';
import { Modal, Checkbox, BasicInput, Dropdown } from '../../index';
import { strings } from '../../../utils';
import { identifiers } from '../../../constants';
import './createModal.css';

const plus = require('../../../assets/images/plus.svg');

class CreateModal extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      createParametersState: this.initializeCreateParameters(),
      showError: false,
      createPassword: true,
    };
  }

  initializeCreateParameters = () => {
    let createParametersState = {};
    this.props.createParameters.forEach(item => {
      if (item.defaultValue) {
        createParametersState[item.id] = item.defaultValue;
      } else if (item.options && item.options.length > 0) {
        createParametersState[item.id] = item.options[0].key;
      }
      else {
        createParametersState[item.id] = item.type === 'boolean' ? false : '';
      }
    });
    return createParametersState;
  }

  setCreateParameters = () => {
    this.setState({ createParametersState: this.initializeCreateParameters(), showError: false });
  }

  create = async () => {
    this.setState({ showError: true });
    if (this.state.createPassword) {
      return this.props.create({ ...this.state.createParametersState, password: undefined }, this.state.createPassword);
    }
    return this.props.create(this.state.createParametersState, this.state.createPassword);
  }

  handleChange = event => {
    this.setState({ createParametersState: { ...this.state.createParametersState, [event.target.id.replace('modal-', '')]: event.target.type === 'checkbox' ? event.target.checked : event.target.value } });
  };

  handleCreatePassword = event => {
    this.setState({ createPassword: !this.state.createPassword });
  }

  renderInput = item => {
    if (item.type === 'boolean') {
      return <Checkbox key={item.id} id={`modal-${item.id}`} text={item.label} value={this.state.createParametersState[item.id]} handleChange={this.handleChange} isDisabled={this.props.isPending} />;
    } else if (item.type === 'select') {
      return <Dropdown key={item.id} id={`modal-${item.id}`} label={item.label} value={this.state.createParametersState[item.id]} handleChange={this.handleChange} options={item.options} />;
    } else if (item.type === 'password') {
      return (
        <React.Fragment key={item.id}>
          <BasicInput id={`modal-${item.id}`} label={item.label} value={this.state.createParametersState[item.id]} handleChange={this.handleChange} type={item.type} isDisabled={this.props.isPending || this.state.createPassword} />
          <Checkbox id={identifiers.USER_HAS_TO_SET_PASSWORD} text={strings.USER_HAS_TO_SET_PASSWORD} value={this.state.createPassword} handleChange={this.handleCreatePassword} isDisabled={this.props.isPending} />
        </React.Fragment>
      );
    }
    return <BasicInput key={item.id} id={`modal-${item.id}`} label={item.label} value={this.state.createParametersState[item.id]} handleChange={this.handleChange} type={item.type} isDisabled={this.props.isPending} />;
  }

  render() {
    return (
      <Modal
        id={strings.CREATE}
        modalButtonText={this.props.primaryButtonText}
        secondaryButtonText={strings.CANCEL}
        handlePrimaryButton={this.create}
        primaryButtonText={strings.SUBMIT}
        primaryButtonClassName="btn-success"
        modalButtonClassName="btn-success btn-create"
        icon={plus}
        title={this.props.title}
        hasHeader
        isPending={this.props.isPending}
        handleModalButton={this.setCreateParameters}
      >
        {this.props.isError && this.state.showError && <div className="alert alert-danger" role="alert">{this.props.errorMessage}</div>}
        <div className="create-modal-input">
          {this.state.createParametersState && this.props.createParameters.map(item => this.renderInput(item))}
        </div>
      </Modal>
    );
  }
}

CreateModal.propTypes = {
  primaryButtonText: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  createParameters: PropTypes.array.isRequired,
  create: PropTypes.func.isRequired,
  isPending: PropTypes.bool,
  isError: PropTypes.bool,
  errorMessage: PropTypes.string,
};

CreateModal.defaultProps = {
  isPending: false,
  isError: false,
  errorMessage: '',
};

export default CreateModal;
