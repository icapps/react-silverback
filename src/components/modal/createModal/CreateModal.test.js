import React from 'react';
import { shallow, configure } from 'enzyme';
import CreateModal from './CreateModal';
import Adapter from 'enzyme-adapter-react-16';


configure({ adapter: new Adapter() });

describe('CreateModal Component', () => {
  it('should render a CreateModal component', () => {
    const wrapper = shallow(
      <CreateModal
        primaryButtonText="test"
        title="test"
        createParameters={[
          { id: 'email', label: 'email', type: "text" },
          { id: 'bool', label: 'bool', type: "boolean" },
          { id: 'password', label: 'password', type: "password" },
        ]}
        create={() => { }}
        isPending={false}
        isError={false}
        errorMessage=""
      />
    );
    wrapper.find('#modal-email').simulate('change', { target: { value: 'test' } });
    wrapper.find('#userHasToSetPassword').simulate('change');
    expect(wrapper).toMatchSnapshot();
  });
});
