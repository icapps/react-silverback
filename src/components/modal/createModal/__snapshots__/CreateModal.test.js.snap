// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateModal Component should render a CreateModal component 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CreateModal
    create={[Function]}
    createParameters={
        Array [
            Object {
              "id": "email",
              "label": "email",
              "type": "text",
            },
            Object {
              "id": "bool",
              "label": "bool",
              "type": "boolean",
            },
            Object {
              "id": "password",
              "label": "password",
              "type": "password",
            },
          ]
    }
    errorMessage=""
    isError={false}
    isPending={false}
    primaryButtonText="test"
    title="test"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        false,
        <div
          className="create-modal-input"
>
          <TextInput
                    errorMessage=""
                    handleChange={[Function]}
                    handleEnter={[Function]}
                    id="modal-email"
                    isDisabled={false}
                    isValid={true}
                    label="email"
                    placeholder=""
                    type="text"
                    value=""
          />
          <Checkbox
                    handleChange={[Function]}
                    id="modal-bool"
                    isDisabled={false}
                    text="bool"
                    value={false}
          />
          <Unknown>
                    <Checkbox
                              handleChange={[Function]}
                              id="userHasToSetPassword"
                              isDisabled={false}
                              text="User has to set the password"
                              value={true}
                    />
                    <TextInput
                              errorMessage=""
                              handleChange={[Function]}
                              handleEnter={[Function]}
                              id="modal-password"
                              isDisabled={true}
                              isValid={true}
                              label="password"
                              placeholder=""
                              type="password"
                              value=""
                    />
          </Unknown>
</div>,
      ],
      "handleModalButton": [Function],
      "handlePrimaryButton": [Function],
      "handleSecondaryButton": [Function],
      "hasHeader": true,
      "icon": "plus.svg",
      "id": "Create",
      "isPending": false,
      "modalButtonClassName": "btn-success btn-create",
      "modalButtonText": "test",
      "primaryButtonClassName": "btn-success",
      "primaryButtonText": "Submit",
      "secondaryButtonClassName": "btn-light",
      "secondaryButtonText": "Cancel",
      "title": "test",
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <TextInput
              errorMessage=""
              handleChange={[Function]}
              handleEnter={[Function]}
              id="modal-email"
              isDisabled={false}
              isValid={true}
              label="email"
              placeholder=""
              type="text"
              value=""
/>,
            <Checkbox
              handleChange={[Function]}
              id="modal-bool"
              isDisabled={false}
              text="bool"
              value={false}
/>,
            <Unknown>
              <Checkbox
                            handleChange={[Function]}
                            id="userHasToSetPassword"
                            isDisabled={false}
                            text="User has to set the password"
                            value={true}
              />
              <TextInput
                            errorMessage=""
                            handleChange={[Function]}
                            handleEnter={[Function]}
                            id="modal-password"
                            isDisabled={true}
                            isValid={true}
                            label="password"
                            placeholder=""
                            type="password"
                            value=""
              />
</Unknown>,
          ],
          "className": "create-modal-input",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "email",
            "nodeType": "class",
            "props": Object {
              "errorMessage": "",
              "handleChange": [Function],
              "handleEnter": [Function],
              "id": "modal-email",
              "isDisabled": false,
              "isValid": true,
              "label": "email",
              "placeholder": "",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "bool",
            "nodeType": "function",
            "props": Object {
              "handleChange": [Function],
              "id": "modal-bool",
              "isDisabled": false,
              "text": "bool",
              "value": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "password",
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Checkbox
                  handleChange={[Function]}
                  id="userHasToSetPassword"
                  isDisabled={false}
                  text="User has to set the password"
                  value={true}
/>,
                <TextInput
                  errorMessage=""
                  handleChange={[Function]}
                  handleEnter={[Function]}
                  id="modal-password"
                  isDisabled={true}
                  isValid={true}
                  label="password"
                  placeholder=""
                  type="password"
                  value=""
/>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "handleChange": [Function],
                  "id": "userHasToSetPassword",
                  "isDisabled": false,
                  "text": "User has to set the password",
                  "value": true,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "errorMessage": "",
                  "handleChange": [Function],
                  "handleEnter": [Function],
                  "id": "modal-password",
                  "isDisabled": true,
                  "isValid": true,
                  "label": "password",
                  "placeholder": "",
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": Symbol(react.fragment),
          },
        ],
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          false,
          <div
            className="create-modal-input"
>
            <TextInput
                        errorMessage=""
                        handleChange={[Function]}
                        handleEnter={[Function]}
                        id="modal-email"
                        isDisabled={false}
                        isValid={true}
                        label="email"
                        placeholder=""
                        type="text"
                        value=""
            />
            <Checkbox
                        handleChange={[Function]}
                        id="modal-bool"
                        isDisabled={false}
                        text="bool"
                        value={false}
            />
            <Unknown>
                        <Checkbox
                                    handleChange={[Function]}
                                    id="userHasToSetPassword"
                                    isDisabled={false}
                                    text="User has to set the password"
                                    value={true}
                        />
                        <TextInput
                                    errorMessage=""
                                    handleChange={[Function]}
                                    handleEnter={[Function]}
                                    id="modal-password"
                                    isDisabled={true}
                                    isValid={true}
                                    label="password"
                                    placeholder=""
                                    type="password"
                                    value=""
                        />
            </Unknown>
</div>,
        ],
        "handleModalButton": [Function],
        "handlePrimaryButton": [Function],
        "handleSecondaryButton": [Function],
        "hasHeader": true,
        "icon": "plus.svg",
        "id": "Create",
        "isPending": false,
        "modalButtonClassName": "btn-success btn-create",
        "modalButtonText": "test",
        "primaryButtonClassName": "btn-success",
        "primaryButtonText": "Submit",
        "secondaryButtonClassName": "btn-light",
        "secondaryButtonText": "Cancel",
        "title": "test",
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <TextInput
                errorMessage=""
                handleChange={[Function]}
                handleEnter={[Function]}
                id="modal-email"
                isDisabled={false}
                isValid={true}
                label="email"
                placeholder=""
                type="text"
                value=""
/>,
              <Checkbox
                handleChange={[Function]}
                id="modal-bool"
                isDisabled={false}
                text="bool"
                value={false}
/>,
              <Unknown>
                <Checkbox
                                handleChange={[Function]}
                                id="userHasToSetPassword"
                                isDisabled={false}
                                text="User has to set the password"
                                value={true}
                />
                <TextInput
                                errorMessage=""
                                handleChange={[Function]}
                                handleEnter={[Function]}
                                id="modal-password"
                                isDisabled={true}
                                isValid={true}
                                label="password"
                                placeholder=""
                                type="password"
                                value=""
                />
</Unknown>,
            ],
            "className": "create-modal-input",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "email",
              "nodeType": "class",
              "props": Object {
                "errorMessage": "",
                "handleChange": [Function],
                "handleEnter": [Function],
                "id": "modal-email",
                "isDisabled": false,
                "isValid": true,
                "label": "email",
                "placeholder": "",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "bool",
              "nodeType": "function",
              "props": Object {
                "handleChange": [Function],
                "id": "modal-bool",
                "isDisabled": false,
                "text": "bool",
                "value": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "password",
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <Checkbox
                    handleChange={[Function]}
                    id="userHasToSetPassword"
                    isDisabled={false}
                    text="User has to set the password"
                    value={true}
/>,
                  <TextInput
                    errorMessage=""
                    handleChange={[Function]}
                    handleEnter={[Function]}
                    id="modal-password"
                    isDisabled={true}
                    isValid={true}
                    label="password"
                    placeholder=""
                    type="password"
                    value=""
/>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "handleChange": [Function],
                    "id": "userHasToSetPassword",
                    "isDisabled": false,
                    "text": "User has to set the password",
                    "value": true,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "errorMessage": "",
                    "handleChange": [Function],
                    "handleEnter": [Function],
                    "id": "modal-password",
                    "isDisabled": true,
                    "isValid": true,
                    "label": "password",
                    "placeholder": "",
                    "type": "password",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": Symbol(react.fragment),
            },
          ],
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
